### PR TITLE
bugfix: color scheme instance mask, refresh with dataset, color picker UI and etc

### PR DESCRIPTION
**Pull request by lanzhenw**
This bugfix fixes the following bugs: 

- [x] On teams and in the UI toggling between different tabs, there was a regression about applying dataset default setting. 
- [x]  instance mask should update with bounding boxes as settings changed
- [x] embeddedDoc filter item color should update
- [x] When user changes datasets in the App, the session.color_scheme needs to update
- [x] JSON editor: colorByAttribute should not be null when using default "label" setting
- [x] Line spacing
- [x] Color picker: simple picker values should sync; fancy color picker had slide and drag issue with selection. 

Instance Mask:
https://github.com/voxel51/fiftyone/assets/17770824/0b3bdad1-802b-40a2-8fc5-7a03b2a9e494

Changing dataset on UI updates session.color_scheme:
https://github.com/voxel51/fiftyone/assets/17770824/852fbfc1-f623-4e25-8454-93307bad9e0e

Chrome color picker:
https://github.com/voxel51/fiftyone/assets/17770824/abe6437d-07f3-437d-ab6c-05199c70bcc0

## What changes are proposed in this pull request?

(Please fill in changes proposed in this fix)

## How is this patch tested? If it is not, please explain why.

(Details)

### Is this a user-facing change that should be mentioned in the release notes?

-   [x] No. You can skip the rest of this section.
-   [ ] Yes. Give a description of this change to be included in the release
        notes for FiftyOne users.

(Details in 1-2 sentences. You can just refer to another PR with a description
if this PR is part of a larger change.)

### What areas of FiftyOne does this PR affect?

-   [x] App: FiftyOne application changes
-   [ ] Build: Build and test infrastructure changes
-   [ ] Core: Core `fiftyone` Python library changes
-   [ ] Documentation: FiftyOne documentation changes
-   [ ] Other

### CODE REVIEW

**Review by manivoxel51 — COMMENTED** [review]
🍨 

question: Is there a way to refactor all the outside click logic into its own hook or re-use and existing one?

File: app/packages/core/src/components/ColorModal/FieldSetting.tsx [hunk 1]
```diff
@@ -40,7 +40,8 @@ type State = {
 };
 
 const FieldSetting: React.FC<Prop> = ({ prop }) => {
-  const colorContainer: React.RefObject<HTMLDivElement> = React.createRef();
+  const wrapperRef: React.RefObject<HTMLDivElement> = React.createRef();
+  const pickerRef: React.RefObject<TwitterPicker> = React.createRef();
```
Comment (sashankaryal) [inline app/packages/core/src/components/ColorModal/FieldSetting.tsx:44]: Why not `useRef()`? Is it intentional that you want these refs to not be memoized between renders?

File: app/packages/core/src/components/ColorModal/FieldSetting.tsx [hunk 2]
```diff
@@ -155,6 +160,21 @@ const FieldSetting: React.FC<Prop> = ({ prop }) => {
     // eslint-disable-next-line react-hooks/exhaustive-deps
   }, [useRecoilValue(resetColor)]);
 
+  useEffect(() => {
+    function handleClickOutside(event) {
+      if (wrapperRef.current && !wrapperRef.current.contains(event.target)) {
+        setShowFieldPicker(false);
+      }
+    }
```
Comment (sashankaryal) [inline app/packages/core/src/components/ColorModal/FieldSetting.tsx:168]: Could you use `useOutsideClick()` hook here?

File: app/packages/core/src/components/ColorModal/colorPalette/ColorPalette.tsx [hunk 3]
```diff
@@ -74,10 +74,30 @@ const ColorPalette: React.FC<ColorPaletteProps> = ({
...
+      }
+    }
+
+    // Bind the event listener
+    document.addEventListener("mousedown", handleClickOutside);
+    return () => {
+      // Unbind the event listener on clean up
+      document.removeEventListener("mousedown", handleClickOutside);
+    };
+  }, [wrapperRef]);
```
Comment (sashankaryal) [inline app/packages/core/src/components/ColorModal/colorPalette/ColorPalette.tsx:91]: Same comment as above, why not `useOutsideClick()`?

File: app/packages/looker/src/lookers/shared.ts [hunk 4]
```diff
@@ -21,10 +23,17 @@ export const LookerUtils = {
     next: Readonly<Optional<BaseState["options"]>>
   ): boolean => {
     let reloadSample = false;
+
     if (next.coloring && current.coloring.seed !== next.coloring.seed) {
       reloadSample = true;
     } else if (next.coloring && next.coloring.by !== current.coloring.by) {
       reloadSample = true;
+    } else if (!_.isEmpty(_.xor(next.coloring.pool, current.coloring.pool))) {
```
Comment (sashankaryal) [inline app/packages/looker/src/lookers/shared.ts:31]: cool

File: app/packages/looker/src/util.ts [hunk 5]
```diff
@@ -547,3 +548,31 @@ export const getMimeType = (sample: any) => {
...
+function compareObjectArr(arr1: object[], arr2: object[]): boolean {
+  const sortedArr1 = arr1
+    .map((el) => JSON.stringify(el, Object.keys(el).sort()))
+    .sort();
+  const sortedArr2 = arr2
+    .map((el) => JSON.stringify(el, Object.keys(el).sort()))
+    .sort();
+
+  return JSON.stringify(sortedArr1) === JSON.stringify(sortedArr2);
+}
```
Comment (sashankaryal) [inline app/packages/looker/src/util.ts:578]: nit: use lodash? something like...
```
function compareObjectArr(arr1: object[], arr2: object[]): boolean {
  const sortedArr1 = sortBy(arr1, (el) => JSON.stringify(el, Object.keys(el).sort()));
  const sortedArr2 = sortBy(arr2, (el) => JSON.stringify(el, Object.keys(el).sort()));

  return isEqual(sortedArr1, sortedArr2);
}
```

File: app/packages/state/src/recoil/atoms.ts [hunk 6]
```diff
@@ -410,3 +410,8 @@ export const sessionColorScheme = atom<ColorSchemeSetting>({
     fields: [],
   },
 });
+
+export const isUsingSessionColorScheme = atom<boolean>({
+  key: "isUsingSessionColorScheme",
+  default: false,
```
Comment (sashankaryal) [inline app/packages/state/src/recoil/atoms.ts:416]: Is this something we'd want to persist in browser storage?

**Review by sashankaryal — CHANGES_REQUESTED** [review]
Some minor comments, but looks good overall!

Requesting changes only because I'm not sure if I see color options for different targets for segmentations the way it's shown in the video. Will clarify offline.

**Comment by sashankaryal** [comment]
> 🍨
> 
> question: Is there a way to refactor all the outside click logic into its own hook or re-use and existing one?

We already have a hook for that, it's called `useOutsideClick()`.

File: app/packages/state/src/recoil/atoms.ts (hunk 6 above)
Comment (lanzhenw) [reply app/packages/state/src/recoil/atoms.ts:416]: No, this is an indicator to determine whether to apply `dataset.appConfig.colorScheme` in loading. (when isUsingSessionColorScheme is false, apply dataset setting). It's set when we load the session state in `useStateUpdate`.

File: app/packages/core/src/components/ColorModal/FieldSetting.tsx (hunk 1 above)
Comment (lanzhenw) [reply app/packages/core/src/components/ColorModal/FieldSetting.tsx:44]: changing to useRef()

**Review by sashankaryal — APPROVED** [review]
LGTM! 🥳

**Pull request merged**